src/types/DataProvider.ts

```diff
@@ -1,15 +1,20 @@
 import { ajv } from '../ajv';
+import { KeycloakId, keycloakIdSchema } from './KeycloakUserId';
 import type { JSONSchemaType } from 'ajv';
-import type { Writable } from './Writable';
 import type { ShortCode } from './ShortCode';
 
 interface DataProvider {
 	readonly shortCode: ShortCode;
 	name: string;
+	keycloakOrganizationId?: KeycloakId | null;
```

Comment:
As discussed [elsewhere](https://github.com/PhilanthropyDataCommons/service/pull/1378#discussion_r1881251198), why make this key optional? It definitely should be nullable! But when should it be missing?

Reply:
One advantage to making `keycloakOrganizationId` optional is it truly is optional when creating a data provider. The way I  implemented "optional" on the relational database side of things is a nullable column on the existing data providers entity. In the API, both the API of the TypeScript code and the HTTP API, we require one less thing from callers that create a data provider. This also means the existing tests become regression tests and new tests that supply the optional parameter can test the optional parameter.

I suppose we could instead add a table to the database that relates PDC Keycloak IDs to PDC service data providers. That would be a more faithful database representation of "optional." The reason I did not do that is the anticipation of dealing with three more tables and the associated joins that result. Perhaps this is the wrong reason to pre-emptively denormalize with a nullable column. If we add the separate tables and make both the PDC Keycloak ID and PDC service data provider/funder/changemaker short code/ID non-null, would we make the `keycloakOrganizationId` optional but not null? Would that satisfy?

Or, I suppose, we could force API callers to include the `keycloakOrganizationId` every time, even if it is null. Would you expect this in the case of using a separate table for the optional attribute?

Reply:
I don't think Jason is asking about database implementation, but rather API design.

Basically: is it more clear to have an attribute that is sometimes there and sometimes isn't, as opposed to a value that is always there and can sometimes be null.

Reply:
@bickelj yes, as @slifty says, I agree with your database design! My question is: when should we accept a `DataProvider` object that does not have a `keycloakOrganizationId` instead of an (logically equivalent) object that has `keycloakOrganizationId` set to `null`:

```ts
let missing: DataProvider = {
  shortCode: s,
  name: "missing KC org id",
  createdAt: "now",
};

let nulled: DataProvider = {
  shortCode: s,
  name: "null KC org id",
  createdAt: "now",
  keycloakOrganizationId: null,
};
```

I believe the only place we're constructing a `DataProvider` in TypeScript code (as opposed to SQL) is the `putDataProvider` handler call to `createOrUpdateDataProvider`, and in that instance there is always a (possibly null) `keycloakOrganizationId` key.

Reply:
@jasonaowen I wasn't sure whether you were offering a preference or pointing out an inconsistency in what I pushed, thus the digression about the database implementation. Regardless, what I pushed did not actually work. Pushing served its purpose of getting good feedback such that I didn't need to re-implement this three times across three entities. I agree that it is simpler (and better) to consistently include the key every time everywhere and let it be `null` and I've taken that approach in recent pushes.